### ISSUE TITLE
Support Dagger's @Reusable annotation

### ISSUE THREAD

**Issue by kirich1409** [yandex/yatagan#19]
I am migrating project from Google Dagger (Dagger 2) to Yatagan and we have a lot of usages of `@Reusable` annotation. It will be good to have an alternative in Yatagan

**Comment by Jeffset** [comment]
Thanks for filing the issue!

Let's see what Dagger's doc say on `@Reusable`:
> A scope that indicates that the object returned by a binding may be (but might not be) reused.
`@Reusable` is useful when you want to limit the number of provisions of a type, but there is no specific lifetime over which there must be only one instance.

So it's only a way to tell the framework, that it's preferable for the type to be constructed less times, then it is requested. But the framework doesn't provide any guarantees on whether it's honored at all. So the program should not depend on any particular behavior imposed by `@Reusable` annotation.

Actually, I find it very difficult to correctly interpret such API's purpose. It might be reasonable to think, that such such nodes are actually **cached** (probably without double-checked locking), but are _compatible with any scope_. There's good answer in SO concerning `@Reusable` - https://stackoverflow.com/a/39154297/8896406

I think we can actually introduce `@Reusable` to Yatagan's API no matter what, as even if it's not implemented at all, it still fits the contract from the doc - it's just that framework may never choose to reuse the instance. I'm going to investigate what Dagger _actually_ does with `@Reusable` nodes and come back with the follow up.

TL;DR: We'll introduce `@Reusable` in Yatagan anyway, though it's concrete behavior is yet to be finalized.

**Comment by Jeffset** [comment]
It seems that `@Reusable` effectively works like **scoped**, except it is implicitly compatible with every scope. Also Dagger uses `SingleCheck.provider()` instead of double checking one.

Worth noting that Yatagan allows bindings (provisions, inject-constructed-classes) to be annotated with multiple scopes to make them compatible with arbitrary number of components.

So, it makes sense for Yatagan to support this as is.
